### ISSUE TITLE
Slicing Dataframe leads only 1 core being use

### ISSUE THREAD

**Issue by thoughtfuldata** [modin-project/modin#3657]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 : 10.0.19041
- **Modin version** (`modin.__version__`): 0.11.2
- **Python version**: 3.9.6
- **Code we can use to reproduce**:

df['name'] = transcript_df[:500].apply(lambda row: my_func(row), axis = 1)

### Describe the problem
If I use the complete dataframe then everything runs correctly and multiple process(PID) run, however if I slice to anything below 250,000 only 1 core is used,

**Comment by dchigarev** [comment]
@thoughtfuldata Thanks for the report!

Currently, Modin does not normalize partitioning after masking a DataFrame, this is done so because repartitioning requires data shuffling between workers, which gives a decent overhead since there is no good implementation in Modin for this yet, modin-core is working on it though. In certain cases overhead of repartitioning is undesired, but in cases like the reported one, it would give a profit.

@devin-petersohn  do you think it would make sense to add a switch that will enable an ability to always normalize partitioning when it was changed (could be useful for `concat` and masking), or do you think we could fix this in general by introducing a proper shuffling mechanism soon?

**Comment by devin-petersohn** [comment]
@dchigarev Yes, we should implement a way to normalize partitioning. It's a bit of an expensive operation, but it would make sense in a case like this.

**Comment by thoughtfuldata** [comment]
@dchigarev Thank you for the response, I look forward to being able to use it in my work
